src/array/merge.ts

```diff
@@ -13,29 +13,26 @@
  * ```
  */
 export function merge<T>(
-  root: readonly T[],
-  others: readonly T[],
-  matcher: (item: T) => any,
+  prev: readonly T[],
+  array: readonly T[],
+  toKey: (item: T) => any,
 ): T[] {
-  if (!others && !root) {
+  if (!array && !prev) {
     return []
   }
-  if (!others) {
-    return [...root]
+  if (!array) {
+    return [...prev]
   }
-  if (!root) {
+  if (!prev) {
     return []
   }
-  if (!matcher) {
-    return [...root]
+  if (!toKey) {
+    return [...prev]
   }
-  return root.reduce((acc, r) => {
-    const matched = others.find(o => matcher(r) === matcher(o))
-    if (matched) {
-      acc.push(matched)
-    } else {
-      acc.push(r)
-    }
-    return acc
+  const keys = array.map(toKey)
+  return prev.reduce((next, prevItem) => {
+    const index = keys.indexOf(toKey(prevItem))
```

Comment:
Now we have loop inside loop, which is `O(N * K)` complexity (here `N = prev.length`, `K = array.length`)

We can achieve a `O(N + K)`  by using `Map` for keys. The problem here that it could be overhead for small arrays. Everything really depends on benchmarks results. lodash has some heuristics for such cases -- they compare the length of the array and select the approach based on that info. Simple loops for short arrays and `Map` for the long ones. But that complexity maybe is not a good fit for simplicity philosophy.

Reply:
> We can achieve a `O(N + K)` by using `Map` for keys.

Interesting, thanks!

> The problem here that it could be overhead for small arrays. Everything really depends on benchmarks results. lodash has some heuristics for such cases -- they compare the length of the array and select the approach based on that info. Simple loops for short arrays and `Map` for the long ones. But that complexity maybe is not a good fit for simplicity philosophy.

We do have benchmarks now, and we're using [Codspeed](https://codspeed.io/) for continuous benchmarking and regression reports. It's currently disabled though, as I need to fix a configuration issue.

I appreciate you bringing up the question on philosophy. I think the general approach with Radash upstream was to prefer brevity over performance. As you can tell, I'm leaning closer to performance than Ray was, so that opens the question of “where do we draw the line?” as you've just pointed out.

As I see it, there are 3 fundamental questions in the context of performance vs brevity:
1. Is this function likely to be in a hot path of a significant percentage of applications?
2. How likely is it to be a bottleneck when it's in a hot path?
3. How much added code is necessary to achieve a significant performance boost?

So it's a case-by-case issue, I believe. Please let me know if you have anything to add on this, as I'm not super experienced in this area.

Reply:
I agree with your points, and I also want to stay away from lodash-like implementation with two algorithms inside the function. The question is do we want to stay with the current solution or try the one with a `Map`? The code with `Map` approach should also be simple but bring some trade-offs. The best way is to implement and benchmark it. I can try to do it later and create a new PR for discussion. Anyway, current changes only improve the old behavior, so let's merge this PR!